### PR TITLE
fix(leios): vote and EB certificate for stake-based committee

### LINKED ISSUE #1782: leios: update vote and EB certificate CBOR for stake-based committee
## Summary

CIP-0164 was updated by cardano-foundation/CIPs#1196, merged on 2026-05-27, to replace the wFA^LS voting committee with stake-based committee selection. gouroboros still models the older certificate shape in `ledger/common/certs.go`: `LeiosEbCertificate` carries `ElectionId`, `PersistentVoters`, `NonpersistentVoters`, and `AggregatedVoteSig`.

The current CIP shape is a uniform vote plus a compact certificate with a committee bitfield and one aggregate BLS signature.

## What Changed

- No per-EB sortition.
- No non-persistent voters.
- No per-vote eligibility proof.
- Votes are uniform: `slot_no`, `endorser_block_hash`, `voter_id`, `vote_signature`.
- Certificates carry `slot_no`, `endorser_block_hash`, `signers` bitfield, and `aggregate_signature`.
- `signers` is a MSB-first bitfield over the epoch committee.
- Leios uses BLS12-381 MinSig, so signatures are 48 bytes.
- Vote offer/request identifiers should identify `(slot, voter_id)` for the epoch committee, not an opaque byte `VoteIssuerId` tied to the old scheme.

## What Needs to Happen

1. Add or replace Leios vote and certificate types to match the current CIP-0164 CDDL.
2. Replace the wFA^LS PV/NPV fields in `LeiosEbCertificate` with the bitfield-based certificate shape.
3. Update LeiosNotify and LeiosFetch vote offer/request structs to use the current vote identifier shape.
4. Decide whether `MsgVotes` should expose typed votes directly or keep raw CBOR plus typed decode helpers.
5. Add CBOR round-trip tests for votes, certificates, empty signer bitfields, and multi-byte signer bitfields.
6. Validate signature lengths and committee-size-dependent bitfield lengths.
7. Update tests and docs that mention persistent/non-persistent voters or sortition proofs.

## Out of Scope

- Stake snapshot computation and quorum validation in a running node.
- Dingo vote storage, aggregation policy, and RB validation wiring.
- Full BLS aggregate verification if that is split into a separate crypto task.

## References

- CIP PR: https://github.com/cardano-foundation/CIPs/pull/1196
- CIP-0164: https://github.com/cardano-foundation/CIPs/blob/master/CIP-0164/README.md
- Existing broad vote diffusion tracker: #1678

### PR DESCRIPTION
**Pull request by wolf31o2**
Closes #1782 

---
## Summary by cubic
Aligns Leios votes and EB certificate with CIP-0164’s stake-based committee. Uses `(SlotNo, VoterId)`, a signer bitfield, and strict 48-byte BLS signatures; updates `protocol/leiosfetch`, `protocol/leiosnotify`, `protocol/leiosvotes`, and clarifies Dijkstra’s placeholder cert slots.

- **Refactors**
  - `ledger/common`: added `LeiosVoteId`, `LeiosVote`, `LeiosEbCertificate`, `LeiosBlsSignatureSize`, signer bitfield helpers (`LeiosSignerBitfieldSize`, `ValidateLeiosSignerBitfield`, `LeiosSignerBit`), and signature validation; EB payload now `{SlotNo, EndorserBlockHash, Signers, AggregatedSignature}` with CBOR, `Validate(committeeSize)`, `Signer(voterId)`, and `Get/SetAggregatedSignature`.
  - `ledger/common/certs`: removed legacy `CertificateTypeLeiosEb` and wrapper handling.
  - `ledger/dijkstra`: documented and tested that `leios_cert`/`peras_cert` remain empty-list placeholders per current CDDL; EB payload is separate.
  - `protocol/leiosfetch`: vote IDs are `(SlotNo, VoterId)`; added `NewMsgVotesFromVotes` and `MsgVotes.DecodeVotes`.
  - `protocol/leiosnotify`: `VotesOffer` now advertises `(SlotNo, VoterId)`.
  - `protocol/leiosvotes`: `Vote` now aliases `ledger/common.LeiosVote`; docs note `Blake2b256` hash and 48-byte BLS.

- **Migration**
  - Replace `ElectionId` with `SlotNo`; replace `PersistentVoters`/`NonpersistentVoters` with `Signers` bitfield and validate size and unused bits.
  - Rename `AggregatedVoteSig` to `AggregatedSignature`.
  - Update `MsgVotesRequest`/`MsgVotesOffer` to `{SlotNo, VoterId}`.
  - Enforce 48-byte BLS signatures via `LeiosVote.Validate()` or typed encode/decode helpers.
  - Note: Dijkstra block cert slots remain CDDL placeholders; use the standalone EB payload struct for CIP-0164 data.

<sup>Written for commit 456fa0e52850f48564b4eeb18e5f9b5639c58dba. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/blinklabs-io/gouroboros/pull/1787?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Stronger vote/certificate validation including aggregated-signature and signer-bitfield checks
  * Votes identified by slot number and numeric voter ID

* **Bug Fixes**
  * Enforced exact BLS signature size for votes and certificates

* **Refactor**
  * Unified vote and certificate data shapes across protocol modules

* **Tests**
  * Added extensive LEIOS validation and CBOR round‑trip tests

* **Documentation**
  * Updated protocol docs for vote request/response formats

[Review Change Stack](https://app.coderabbit.ai/change-stack/blinklabs-io/gouroboros/pull/1787?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: ledger/common/certs.go [hunk 1]
```diff
@@ -1478,15 +1478,28 @@ func (c *UpdateDrepCertificate) Type() uint {
 type LeiosEbCertificate struct {
 	cbor.StructAsArray
 	cbor.DecodeStoreCbor
-	ElectionId          Blake2b256
+	CertType            uint
```
Comment (chrisguiney) [inline ledger/common/certs.go]: **Blocking: this serializes the certificate as a 5-element array led by the type tag `19`, but CIP-0164 `leios_certificate` is a bare 4-tuple with no discriminator.**

Per [CIP-0164 Appendix B](https://github.com/cardano-foundation/CIPs/blob/master/CIP-0164/README.md):
```cddl
leios_certificate =
  [ slot_no, endorser_block_hash, signers : bytes, aggregated_signature : leios_bls_signature ]
```
and it is embedded directly as the optional 6th element of the block, not via the transaction `certificate` union:
```cddl
ranking_block = [ header, transaction_bodies, …, invalid_transactions, ? eb_certificate : leios_certificate ]
```

Adding `CertType` here and emitting `c.Type()` in `MarshalCBOR` breaks both directions (byte-level repro):
```
conformant 4-tuple:  84 18 63 5820…41 80 5830 bb…    (array(4))
this PR emits:        85 13 18 63 5820…41 80 5830 bb… (array(5), leading 0x13 = 19)

decode a conformant 4-tuple → "cannot decode CBOR array to struct with different number of elements"
```
So we emit an extra `19` a peer rejects, and we can't parse a real `leios_certificate`. Note `LeiosVote` (same `MarshalCBOR` pattern) correctly omits the tag and matches `leios_vote` — only the cert regressed; #1782 also lists the cert as 4 fields.

**Fix:** drop `CertType` from the struct and remove `c.Type()` from the `MarshalCBOR` array so it encodes `[SlotNo, EndorserBlockHash, Signers, AggregatedSignature]`. Keep the `Type()` method (interface requirement) but don't serialize it. Separately, since the EB cert is an RB field rather than a tx-certificate, it shouldn't really flow through `CertificateWrapper`/`CertificateTypeLeiosEb` (pre-existing from #1195/#1240) — worth decoupling as part of CDDL conformance.

File: ledger/common/leios_test.go [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+	})
+	require.NoError(t, err)
+
+	var decoded LeiosVote
+	_, err = cbor.Decode(encoded, &decoded)
+	require.Error(t, err)
+	assert.Contains(t, err.Error(), "VoteSignature")
+}
+
+func TestLeiosEbCertificateCborRoundTrip(t *testing.T) {
```
Comment (chrisguiney) [inline ledger/common/leios_test.go:105]: Every cert test here is PR-encode → PR-decode, so both sides share the same (non-conformant) 5-tuple convention and pass regardless of wire correctness — the exact gap #1782 item 5 calls out. Please add a known-answer test that decodes a hand-built CDDL vector into `LeiosEbCertificate`:
```go
raw, _ := cbor.Encode([]any{uint64(99), NewBlake2b256(...), []byte{0x80}, leiosTestSignature(0xbb)})
var c LeiosEbCertificate
_, err := cbor.Decode(raw, &c) // must succeed and round-trip byte-for-byte
```
That single test pins the wire format against the CIP and would have caught the leading-tag issue. (Worth one for the vote too, even though the vote is currently correct.)

File: protocol/leiosfetch/messages.go [hunk 3]
```diff
@@ -176,6 +173,36 @@ func NewMsgVotes(votes []cbor.RawMessage) *MsgVotes {
...
+}
+
+func (m *MsgVotes) DecodeVotes() ([]lcommon.LeiosVote, error) {
+	ret := make([]lcommon.LeiosVote, 0, len(m.VotesRaw))
+	for idx, voteRaw := range m.VotesRaw {
+		var vote lcommon.LeiosVote
+		if _, err := cbor.Decode(voteRaw, &vote); err != nil {
+			return nil, fmt.Errorf("decode vote %d: %w", idx, err)
+		}
+		if err := vote.Validate(); err != nil {
```
Comment (chrisguiney) [inline protocol/leiosfetch/messages.go]: Redundant: `LeiosVote.UnmarshalCBOR` already calls `Validate()`, so `cbor.Decode` on line 195 fails before this is reached — the explicit `vote.Validate()` here is dead code. Harmless, but can be dropped (or keep it and simplify `UnmarshalCBOR` to not validate, picking one layer).

**Review by chrisguiney — CHANGES_REQUESTED** [review]
**Requesting changes** — the migration is clean and well-tested (vote shape, MSB-first bitfield, and 48-byte BLS sizing all match CIP-0164), but the **EB certificate's CBOR is not CDDL-conformant** and won't interop with a conforming node. Details inline. Once the certificate serializes as the bare 4-tuple, this is good to go.

Verified correct: `leios_vote` shape (4-tuple, no tag) ✓; `signers` MSB-first with `⌈N/8⌉` length + trailing-bit check ✓; 48-byte MinSig enforced on vote + aggregate ✓; `SetAggregatedSignature` clears cached CBOR ✓; `(SlotNo, VoterId)` identifiers ✓; builds + touched-package tests pass.

**Review by chrisguiney — APPROVED** [review]
